module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+	LRUEjection    = EjectionMode("lru-ejection")
+)
+
+// number of entries we try to eject before giving up and ejecting LRU
+const maximumRandomTrials = 10
+
+// EIndex is data type representing an entity index in EntityDoubleLinkedList.
+type EIndex uint32
+
+// doubleLinkedListPointer represents a slice-based linked list pointer. Instead of pointing
+// to a memory address, this pointer points to a slice index.
+//
+// Note: an "undefined" (i.e., nil) notion for this doubleLinkedListPointer corresponds to the
+// value of uint32(0). Hence, legit "pointer" values start from uint32(1).
+// doubleLinkedListPointer also furnished with methods to convert a "pointer" value to
+// a slice index, and reverse.
+type doubleLinkedListPointer struct {
+	pointerValue uint32
+}
+
+// isUndefined returns true if this pointer is set to its zero value. An undefined
+// slice-based pointer is equivalent to a nil address-based one.
+func (d doubleLinkedListPointer) isUndefined() bool {
+	return d.pointerValue == uint32(0)
+}
+
+// setUndefined sets sliced-based pointer to its undefined (i.e., nil equivalent) value.
+func (d *doubleLinkedListPointer) setUndefined() {
+	d.pointerValue = uint32(0)
+}
+
+// sliceIndex returns the slice-index equivalent of the pointer.
+func (d doubleLinkedListPointer) sliceIndex() EIndex {
+	return EIndex(d.pointerValue) - 1
+}
+
+// setPointer converts the input slice-based index into a slice-based pointer and
+// sets the underlying pointer.
+func (d *doubleLinkedListPointer) setPointer(sliceIndex EIndex) {
+	d.pointerValue = uint32(sliceIndex + 1)
+}
+
+// doubleLinkedListNode represents a slice-based double linked list node that
+// consists of a next and previous pointer.
+type doubleLinkedListNode struct {
+	next doubleLinkedListPointer
+	prev doubleLinkedListPointer
+}
+
+// doubleLinkedList represents a double linked list by its head and tail pointers.
+type doubleLinkedList struct {
+	head doubleLinkedListPointer
+	tail doubleLinkedListPointer
+}
+
+func newDoubleLinkedList() *doubleLinkedList {
+	return &doubleLinkedList{
+		head: doubleLinkedListPointer{pointerValue: 0},
+		tail: doubleLinkedListPointer{pointerValue: 0},
+	}
+}
+
+// cachedEntity represents a cached entity that is maintained as a double linked list node.
+type cachedEntity struct {
+	owner  uint64
+	node   doubleLinkedListNode
+	id     flow.Identifier
+	entity flow.Entity
+}
+
+type EntityDoubleLinkedList struct {
+	total        uint32
+	free         *doubleLinkedList // keeps track of free slots.
+	used         *doubleLinkedList // keeps track of allocated slots to cachedEntities.
+	values       []cachedEntity
+	ejectionMode EjectionMode
+}
+
+func NewEntityList(limit uint32, ejectionMode EjectionMode) *EntityDoubleLinkedList {
+	l := &EntityDoubleLinkedList{
+		free:         newDoubleLinkedList(),
+		used:         newDoubleLinkedList(),
+		values:       make([]cachedEntity, limit),
+		ejectionMode: ejectionMode,
```

Comment:
If this code would be abstracted further, it would be great to have a mode which does not feature any ejection / eviction as suggested here [1]. It might be useful to be able to use this abstract code for other algorithms such as MTrie which can make use of applying state to caller supplied arrays without necessarily having an eviction or ejection mode...

[1] https://www.notion.so/dapperlabs/Intro-to-heap-friendly-hero-structures-cd613aba6f4244f88e58fec1f222f476

Reply:
Issued up here: https://github.com/dapperlabs/flow-go/issues/6085

Reply:
Thanks, commented in the issue :-)